Doppler.Sap/Factory/SapServiceSettingsFactory.cs

```diff
@@ -0,0 +1,39 @@
+using Doppler.Sap.Enums;
+using Doppler.Sap.Models;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Doppler.Sap.Factory
+{
+    public class SapServiceSettingsFactory : ISapServiceSettingsFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IOptionsMonitor<SapServiceConfig> _serviceSettings;
+
+        public SapServiceSettingsFactory(IServiceProvider serviceProvider, IOptionsMonitor<SapServiceConfig> serviceSettings)
+        {
+            _serviceProvider = serviceProvider;
+            _serviceSettings = serviceSettings;
+        }
+
+        public ISapTaskHandler CreateHandler(CountryCodeEnum countryCode)
+        {
+            var sapTaskHandler = (ISapTaskHandler)_serviceProvider.GetService(typeof(ISapTaskHandler));
```

Comment:
Why is it not injected in the constructor in place of using the _ServiceProvider_?

Reply:
Oh, because it is transient!

Mmm, I think that it is better to create the instance here than using dependency injection and a method to set the settings.

What do you think about this:

```csharp
        public ISapTaskHandler CreateHandler(CountryCodeEnum countryCode)
        {
            switch (countryCode)
            {
                case CountryCodeEnum.AR:
                    return new SapTaskHandler(_serviceSettings.Get("SapArgService"));
                case CountryCodeEnum.US:
                    return new SapTaskHandler(_serviceSettings.Get("SapUsaService"));
                default:
                    throw new ArgumentException(nameof(countryCode), $"The countryCode '{countryCode}' is not supported.");
            }
```

Reply:
Also, if you change the settings by an array of configurations, you can replace the _switch_ by a search in a list, and you will make adding new countries dynamic.

Reply:
yes, I will work on it.